lifterlms integration

thomasplevy opened the pull request:
Proposed LifterLMS Integration as per issue #1119 

A few things to note:

The current stable version of LifterLMS is 2.6.3 and in that release we added an action hook that we're hooking into to kick the whole thing off. This is `lifterlms_order_complete`, we have a few older actions that we *could* use but nothing that was passing order related information so we could reliably get the reference. We decided to just add an action to keep the amount of querying low inside AffiliateWP.

That said, should we worry about doing a version compare inside AffiliateWP when the integration is enabled? Is there any warning methods available in AffiliateWP to note that version upgrade requirement? We'll, of course, notify our users who are aching to get this built that they need this version to use the integration but I'm concerned about anyone who organically stumbles across the integration and clicks the button.

We've also included some inline JS that would run on LifterLMS Courses & Memberships to make the UI a little more pleasant for our users. This isn't necessary if you're opposed to the inline JS included in this PR. I just wanted to note that in case it looks extremely offensive to you please let me know and we'll rip it out and resubmit. I'm not a huge fan of the inline JS but I am a fan of the UX it provides. I didn't see precedence for adding a whole JS file for the integration (and frankly that doesn't feel necessary if the inline is acceptable to ya'll).


Thanks so much!

pippinsplugins commented on includes/integrations/class-lifterlms.php:
```diff
@@ -0,0 +1,725 @@
+		}
+
+		// if this was a referral or we have a coupon and a coupon affiliate id
+		if ( $this->was_referred() || ( $order->coupon && $order->coupon_affiliate_id ) ) {
```
Let's put this check at the very top. No need to run any logic if the customer wasn't referred.

thomasplevy replied:
👍

pippinsplugins commented on includes/integrations/class-lifterlms.php:
```diff
@@ -0,0 +1,725 @@
+			$affiliate_id = $this->get_affiliate_id( $order_id );
+
+			// use our coupon affiliate if we have one
+			// @note 	is this the correct way to do this? Always override with the coupon?
```
Yes, coupon always takes precedence.

thomasplevy replied:
thanks

pippinsplugins commented on includes/integrations/class-lifterlms.php:
```diff
@@ -0,0 +1,725 @@
+
+			// update existing referral if it exists
+			// this isn't currently ever going to happen with LifterLMS but leaving it here for future use
+			if ( $existing ) {
```
Since $existing isn't defined, it will need to be defined or this logic removed.

thomasplevy replied:
$existing is defined on line 91 (https://github.com/AffiliateWP/AffiliateWP/pull/1257/files#diff-8893037b472760bac3785d4bbd55b49dR91)

pippinsplugins commented:
Looks really good!

I don't mind the inline JS but let's go ahead and move it to our admin JS file so that it's combined with the rest of our JS. Any issues doing that?

thomasplevy commented:
@pippinsplugins no problem re the JS I just didn't want to be so arrogant as to add some JS to one of your files.

Will add to the JS and run a few more tests and update the fork in a little while

pippinsplugins commented:
Great, thanks!

thomasplevy commented:
@pippinsplugins I can't see any obvious minification task in the gruntfile or related directories, how do you handle minification?

thomasplevy commented:
moved inline js out of class-lifterlms.php and fiddled with the JS to ensure nothing gets funky. As noted above I don't see any JS minification tasks in the gruntfile so the JS is not minified, enabled SCRIPT_DEBUG to do some testing on the unminified. Also added a check so that llms related JS will only bind on pages it's needed see the init function.

Please let me know if you see any other issues with the integration and thanks again!

pippinsplugins commented:
We minify before each release so nothing to worry about there!

pippinsplugins commented:
We'll get this reviewed and tested further shortly after our 1.8 release, which is scheduled for Thursday this week.

ramiabraham commented on assets/js/admin.js:
```diff
@@ -228,3 +228,108 @@ jQuery(document).ready(function($) {
+	};
+
+	// instatiate the class
+	var a = new window.llms.metabox_product_affwp();
```
Th AffiliateWP `admin.js` is for items relating to AffiliateWP core, or _all_ integrations, and cannot contain code relating to a single integration.

Will move this to the `class-lifterlms.php` file as inline js for now, however, allowing integrations to define their own assets is something we're considering for a future update.

ramiabraham commented:
@thomasplevy Thanks for your work here! We're going to get this into AffiliateWP `1.8.3`. One adjustment is that we chose to move the js back inline. It's defined in the method `inline_js` within the LMMS integration class.

thomasplevy commented:
@ramiabraham,

I actually think this is better given the current architecture, I hate the idea of bloating (however small) your entire JS file for users who have no interest in LifterLMS.

Thanks for the update and please let me know if you need anything from me!

pippinsplugins commented:
Implemented through https://github.com/AffiliateWP/AffiliateWP/pull/1368
